### PR TITLE
Implement support for multiple Icinga2 Master API URLs

### PR DESCRIPTION
**Pull request by pdorschner**
This PR fixes '[Feature Request] Support for Redundant Master API's' #57 

It' now possible to specify one or more Icinga-URLs. If one URL is not reachable, it'll switch to the URL which doesn't response with an error.
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the README.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Link this PR to related issues.

### CODE REVIEW

**Comment by simu** [comment]
Hi, thanks for the contribution, I'll have a look soon

**Comment by pdorschner** [comment]
Hi,
thank you for your reply. I added a bunch of changes, e.G. Readme-Update, Dockerfile, etc.
More over, the changes/PR has been tested and it worked as expected. So can I help you with something to make this PR get merged?

Best regards

**Comment by pdorschner** [comment]
The Build fails because there are missing methods in the go-icinga2-client, see https://github.com/vshn/go-icinga2-client/pull/9

File: serve.go [hunk 1]
```diff
@@ -102,9 +103,50 @@ func (s *ServeCommand) startHeartbeat() error {
...
+
 		for ts := range s.heartbeatTicker.C {
 			if err := s.heartbeat(ts); err != nil {
 				s.logger.Errorf("sending heartbeat: %s", err)
+
+				// In some rare cases there could be an Icinga2 reload for a config update and the API is not reachable.
+				// So, the switch off of the Config-Master would take in place, but we don't want an unnecessary switch
+				// to the secondary Icinga2 instance.
+				s.logger.Infof("Waiting 5sec for reconnect")
+				time.Sleep(5 * time.Second)
```
Comment (simu) [inline serve.go:124]: having a hard-coded 5 second sleep in a loop with an unknown (configurable at runtime) interval seems risky, although 5 seconds is probably low enough that we'll never run into issues with missing heartbeat updates, as running the heartbeat on an interval < 30 seconds probably doesn't make a lot of sense for most setups.

File: config/config.go [hunk 2]
```diff
@@ -171,15 +171,36 @@ func newIcingaClient(c *SignaliloConfig, l logr.Logger) (icinga2.Client, error)
...
+	var tmp []string
+
+	for _, ur := range c.IcingaConfig.URL {
+		st := strings.Split(ur, ",")
+		for _, b := range st {
+			tmp = append(tmp, b)
+		}
+	}
+
+	c.IcingaConfig.URL = tmp
```
Comment (simu) [inline config/config.go:185]: This is strictly speaking not necessary, as you can provide repeatable arguments in environment variables separated by newlines.

If we really want to keep this, the logic should go into `ConfigInitialize()` as we only need to do this once after Signalilo starts (side-note: `newIcingaClient()` is only called once as well, but logically, it doesn't make sense that `newIcingaClient()` modifies the configuration object).

File: config/config.go [hunk 3]
```diff
@@ -171,15 +171,36 @@ func newIcingaClient(c *SignaliloConfig, l logr.Logger) (icinga2.Client, error)
...
+			return nil, err
+		}
+
+		if err = client.TestIcingaApi(url); err != nil {
+			continue
+		} else {
+			break
+		}
 	}
 	return client, nil
```
Comment (simu) [inline config/config.go:205]: We need to make sure that we have a working client before returning here. Something like

```suggestion
	for _, url := range c.IcingaConfig.URL {
		client, err = icinga2.New(icinga2.WebClient{
			URL:               url,
			Username:          c.IcingaConfig.User,
			Password:          c.IcingaConfig.Password,
			Debug:             c.IcingaConfig.Debug,
			DisableKeepAlives: c.IcingaConfig.DisableKeepAlives,
			TLSConfig:         tlsConfig})
		if err != nil {
			return nil, err
		}

		if err = client.TestIcingaApi(url); err != nil {
			// clear client if the API url wasn't reachable
			client = nil
			continue
		} else {
			break
		}
	}
	if client == nil {
		return nil, fmt.Errorf("No valid Icinga API URL found")
	}
	return client, nil
```

File: config/config.go [hunk 4]
```diff
@@ -171,15 +171,36 @@ func newIcingaClient(c *SignaliloConfig, l logr.Logger) (icinga2.Client, error)
...
+	c.IcingaConfig.URL = tmp
+
+	for _, url := range c.IcingaConfig.URL {
+		client, err = icinga2.New(icinga2.WebClient{
+			URL:               url,
+			Username:          c.IcingaConfig.User,
+			Password:          c.IcingaConfig.Password,
+			Debug:             c.IcingaConfig.Debug,
+			DisableKeepAlives: c.IcingaConfig.DisableKeepAlives,
+			TLSConfig:         tlsConfig})
```
Comment (simu) [inline config/config.go:194]: We don't really need to create a fresh client to test each URL. Calling `SetIcingaUrl()` for the first URL which passes `TestIcingaApi()` would suffice. 

Something like

```
client, err := icinga2.New(icinga2.WebClient{
	URL:               c.IcingaConfig.URL,
	Username:          c.IcingaConfig.User,
	Password:          c.IcingaConfig.Password,
	Debug:             c.IcingaConfig.Debug,
	DisableKeepAlives: c.IcingaConfig.DisableKeepAlives,
	TLSConfig:         tlsConfig})
if err != nil {
	// Failed to initialize the Icinga web client, return error
	return nil, err
}
for _, url := range c.IcingaConfig.URL {
	if err = client.TestIcingaApi(url); err == nil {
		client.SetIcingaUrl(url)
		break
	}
}
if err != nil {
	// We didn't find a valid API URL, return error
	return nil, fmt.Errorf("No valid Icinga API URL found")
}
return client, nil
```

**Review by simu — CHANGES_REQUESTED** [review]
Hi,

I had a closer look (also at the go-icinga2-client PR, see my review there) and found some things that I'd like to have cleaned up before going ahead and merging this PR (after the client library PR).

Overall, the implementation as it is currently does quite a bit of unnecessary work (as pointed out in inline comments).

Additionally, please keep the README.md formatted with one sentence per line (especially the parts where you didn't make any changes) as this makes reviewing changes much easier than wrapped paragraphs.

File: README.md [hunk 5]
```diff
@@ -48,7 +48,9 @@ Mandatory
 * `--icinga_hostname`/`SIGNALILO_ICINGA_HOSTNAME`:
   Name of the Servicehost in Icinga2.
 * `--icinga_url`/`SIGNALILO_ICINGA_URL`:
-  URL of the Icinga API.
+  URL of the Icinga API. It's possible to specify one or more URLs. Via CLI, the parameter can be repeated. If the
+  Parameter is used as ENV, the URLs will be splittet via comma e.G.: `"http://example.com:5665,http://example2.com:5665"`.
```
[the remark this answers is not in the document]
Comment (pdorschner) [reply README.md:52]: That's true, but in my special case, I can't specify multiple URLs with `\n` in my config. Personally

[the remark this answers is not in the document]
Comment (pdorschner) [reply README.md]: > Author

Nevermind, it should be implemented as you said with `\n`

**Comment by simu** [comment]
Can you please update go-icinga2-client to v0.0.16 in this PR? Also, from what I can see, this PR isn't updated yet to use the final variant of the new API methods introduced in https://github.com/vshn/go-icinga2-client/pull/9.

**Comment by pdorschner** [comment]
Hi,
sorry for these inconveniences... now it should be fine

**Review by simu — APPROVED** [review]

**Pull request merged**